[superseded] [os] fix #8734, #8353: parentDir, parentDirs now work reliably

timotheecour opened the pull request:
parentDir, parentDirs now work reliably

* fix #8734
* fix #8353
* removes workaround in ./compiler/packagehandling.nim
> `XXX os's parentDirs is stupid (multiple yields) and triggers an old bug...`

+ many other issues I saw while working on the fix.

I also added extensive unittests so you can now see exactly what `parentDir` and `parentDirs` return

to view the diff and not get confused by code that was only moved, see instructions here https://github.com/nim-lang/Nim/pull/10435/files

## note
the `# todo: fix` is a (commented out) test case for a pre-existing failure from https://github.com/nim-lang/Nim/issues/8395 which i can fix after this PR

Araq commented on lib/pure/os.nim:
```diff
@@ -666,6 +589,86 @@ proc isAbsolute*(path: string): bool {.rtl, noSideEffect, extern: "nos$1".} =
+  ##   * an absolute path remains absolute
+  ##   * trailing ``.`` and ``/`` are resolved before taking the parent dir
+  ## It returns empty when attempting to take parent of ``.`` or ``..``
+  ## to indicate it is invalid and distinguish from ``.``
```
"different from"

timotheecour replied:
done

Araq commented on compiler/packagehandling.nim:
```diff
@@ -7,21 +7,13 @@
 #
 
-iterator myParentDirs(p: string): string =
-  # XXX os's parentDirs is stupid (multiple yields) and triggers an old bug...
```
Your new better parentDirs still has multiple yields and contains logic I don't need.

timotheecour replied:
as discussed here https://irclogs.nim-lang.org/04-02-2019.html#10:41:49 let's ensure correctness first and defer to followup PRS to address that point if we find out it's a performance/code size bottlenecks; there are options (at cost of a bit uglier code)

Araq replied:
The point is: Don't make the compiler's code worse, ``myParentDirs`` still is better than the one in os.nim. Undo the change to compiler/packagehandling.nim

timotheecour replied:
done, reverted; we can fix that 'multiple yield' in subsequent PRs, i can think of 2 ways

Araq commented on tests/stdlib/tos.nim:
```diff
@@ -333,3 +365,102 @@ block ospaths:
+  # fromRoot=false, inclusive=true
+  test parentDirs("a/b/c".unixToNativePath), @["a/b/c", "a/b", "a"]
+  test parentDirs("/a/b/c".unixToNativePath), @["/a/b/c", "/a/b", "/a", "/"]
+  test parentDirs("//a/b//c//".unixToNativePath), @["//a/b//c", "//a/b", "//a", "/"]
```
Why doesn't parentDirs normalize the paths?

timotheecour replied:
same behavior as in most (all?) other languages: it normalizes only the endpoint:

```nim
os.path.dirname("//abc//def//")
Out[3]: '//abc//def'
```

D:
```nim
rdmd --eval 'writeln(dirName("//abc//def//"))'
//abc
```

bash:
```
dirname //abc//def//
//abc
```

etc...

Araq replied:
Sounds like a bug to me, but ok.

timotheecour commented:
> #8734 is arguably not a bug. Especially not if you keep the behaviour that an empty string signals a wrong usage.

#8734 is not 1 bug but multiple bugs and as discussed here https://irclogs.nim-lang.org/04-02-2019.html#10:56:02 and elsewhere in this conversation thread some of the items are indeed bugs, and 1 of them is being debated.
So i added a simple CT switch (that is currently set to `parentDirOfRootIsEmpty = true` ) to unblock this PR and we can address that point later, which'll only require resetting `parentDirOfRootIsEmpty = false`  if we decide to go with that

timotheecour commented:
@Araq PTAL all comments addressed

Araq commented on tests/stdlib/tos.nim:
```diff
@@ -333,3 +365,108 @@ block ospaths:
+    ("/a.txt", "/"),
+  ]
+
+  doAssert runTestCases("parentDir", examples, parentDir(it))
```
That is cargo-cult programming. Now I can look at a list of tuples all day long before I can see it's about ``parentDir``. DRY and meta programming are for creating higher abstractions, not to save keystrokes.

timotheecour replied:
> Now I can look at a list of tuples all day long before I can see it's about `parentDir`

I added documentation to make everything clear. 

I had initially started with the `one doAssert per test case` but that became quickly un-manageable with the number of test cases, in particular:
* the important part of each test case was lost in noise of the repetitive surrounding code, eg `  doAssert parentDir("/usr/local/bin".unixToNativePath) == "/usr/local".unixToNativePath` ; instead, I'm only highlighting the part that changes
* on failure, it doesn't tell you what `parentDir` produced as output (because you don't like `assertEquals` for some reason)
* it stops on 1st error instead of showing all errors, making it terrible when you need to change the implementation of `parentDirs`; instead I'm showing all errors before failing
* if you have to change something, you have to change it everywhere; unlike here where I need to change it in only one place. yes, it's `DRY`, it's not `cargo cult`

## note: the lack of this kind of thorough testing as done here for `parentDir` is the root cause for all those issues and regressions I keep running into when using `std/os` procs (eg https://github.com/nim-lang/Nim/pull/10018 , https://github.com/nim-lang/Nim/pull/10047 , etc)

I have other PR's in the pipeline that fix other bugs in std/os (eg `"/a" /../ "d"`  should be `"/d"`, not `"d"` + other bugs with other procs) so let's please move on with this PR

Araq replied:
> the lack of this kind of thorough testing as done here for `parentDir` is the root cause for all those issues and regressions I keep running into when using `std/os` procs (eg #10018 , #10047 , etc)

Maybe but it's just you. Most people don't use double slashes in all their paths. And your improved implementation adds plenty of its own, new problems.

timotheecour replied:
@araq
* I don't understand your comment. #10018, #10047, #10545 and other PRs I wrote for path handling fix bugs and regressions unrelated to double slashes; incorrect handling of paths with double slashes were just one of the issues I've fixed.
* all languages I've worked with correctly handle paths with double slashes, eg: C++(boost), D, python, java, js(node) etc...; as I already explained
* Double slashes can occur from a variety of reasons, eg:
```
export buildDir=/path/to/build/
export logFile=$buildDir/log.txt # => expands to a path with double slash
```
* see also https://unix.stackexchange.com/questions/212805/should-i-use-a-slash-at-the-end-of-path-variables-in-shell-script-or-not 
> To be on the safe side, include the slash. This can lead to multiple slashes when concatenating the paths, but at least you avoid problems.

and this: https://unix.stackexchange.com/questions/1910/how-does-linux-handle-multiple-consecutive-path-separators-home-username
> Multiple slashes are allowed and are equivalent to a single slash

> And your improved implementation adds plenty of its own, new problems

I'm not sure which are these new problems you're referring to, after this PR changed from `parentDir("/") = "/"` to `""`; which code was previously correct that became incorrect after this PR?

Araq commented on tests/stdlib/tos.nim:
```diff
@@ -26,6 +26,31 @@ Raises
+  Takes care of calling unixToNativePath.
+  Returns true on success.
+  ]##
+  runnableExamples:
```
Who runs ``nim doc tests/stdlib/tos.nim`` in order to check this runnable example? Not me.

timotheecour replied:
fixed, moved to a test in a block

Araq commented on tests/stdlib/tos.nim:
```diff
@@ -26,6 +26,31 @@ Raises
 # test os path creation, iteration, and deletion
 
 import os, strutils, pathnorm
+import "$nim/compiler/unittest_light"
```
unittest is bad because it's full of ``when compiles``. Guess what, unittest_light repeats the mistake. The tests were fine before without this import.

timotheecour replied:
I removed that import for the sake of moving fwd with this PR, but please explain what is wrong with `unittest_light` having `compiles` ?
* assertEquals is written as `if lhs!=rhs: doAssert false, mismatch(lhs, rhs)` so the `compiles` is just used for printing the error msg (inside `mismatch`), not for correctness validation
* theres is already a test for that tests/compiler/tunittest_light.nim which tests `assertEquals` using `doAssert` so if `assertEquals` had bugs, `doAssert` would catch it

If it's a compilation speed issue, there are easy options.

More generally, for `assertEquals`, `unittest.check`, `fmt`, and `sequtils`: 1 argument I heard against more widespread use in testament was "who is guarding the guardian" argument; I'm all for adding more thorough testing (as needed) for these but we shouldn't use `FUD` arguments that they may have bugs that would render testament useless; 
If a bug ever slips through existing checks, it'll eventually be found in the wild, reported, and fixed;
note, outside of testament, these are pretty popular tools for testing in nimble packages, for a reason.

Araq replied:
nim-regex doesn't compile with devel because it uses unittest and unittest does nasty stuff. Yeah, these are compiler regressions, but still, a testing infrastructure shouldn't stress the compiler so much that it becomes hard to debug itself. No FUD here, just lots of experience with unittest.

timotheecour replied:
the concerns you're raising seem specific to `unittest`, not `unittest_light`, as I described here https://github.com/nim-lang/Nim/pull/10545#discussion_r254069436

Araq commented on lib/pure/os.nim:
```diff
@@ -666,6 +589,90 @@ proc isAbsolute*(path: string): bool {.rtl, noSideEffect, extern: "nos$1".} =
+      assert parentDir("foo") == "."
+      assert parentDir("/foo") == "/"
+      assert parentDir(".") == ""
+  const parentDirOfRootIsEmpty = true # still controversial
```
Remove this, it may be controversial but we'll never have ``parentDirOfRootIsEmpty = false``

Araq commented on lib/pure/os.nim:
```diff
@@ -468,7 +388,7 @@ proc splitFile*(path: string): tuple[dir, name, ext: string] {.
   ## Splits a filename into `(dir, name, extension)` tuple.
   ##
-  ## `dir` does not end in `DirSep <#DirSep>`_.
+  ## `dir` does not end in `DirSep <#DirSep>`_ unless it is absolute.
```
Misleading doc: So ... absolute paths can end in the slash? Why? Only true for the root directory "/" or what?

timotheecour replied:
fixed

timotheecour commented:
@Araq comments addressed and green, PTAL

Araq commented on lib/pure/os.nim:
```diff
@@ -666,6 +589,82 @@ proc isAbsolute*(path: string): bool {.rtl, noSideEffect, extern: "nos$1".} =
+      assert parentDir("/") == ""
+  if path == "": return ""
+  result = path
+  while true:
```
Convoluted. What's wrong with fixing ``parentDirPos``?

Araq commented on lib/pure/os.nim:
```diff
@@ -666,6 +589,82 @@ proc isAbsolute*(path: string): bool {.rtl, noSideEffect, extern: "nos$1".} =
+  result = path
+  while true:
+    normalizePathEnd(result)
+    let (dir, name, ext) = splitFile(result)
```
Why would I do that? What has ``ext`` to do with anything?

Araq commented on lib/pure/os.nim:
```diff
@@ -666,6 +589,82 @@ proc isAbsolute*(path: string): bool {.rtl, noSideEffect, extern: "nos$1".} =
+  while true:
+    normalizePathEnd(result)
+    let (dir, name, ext) = splitFile(result)
+    if name == ".." or (name.len == 0 and ext.len == 0): return ""
```
parentDir("bar/baz/foo/..") could just work...

Araq commented on lib/pure/os.nim:
```diff
@@ -666,6 +589,82 @@ proc isAbsolute*(path: string): bool {.rtl, noSideEffect, extern: "nos$1".} =
+    normalizePathEnd(result)
+    let (dir, name, ext) = splitFile(result)
+    if name == ".." or (name.len == 0 and ext.len == 0): return ""
+    if name == ".":
```
Ah so silly single dots are allowed.

Araq commented on lib/pure/os.nim:
```diff
@@ -666,6 +589,82 @@ proc isAbsolute*(path: string): bool {.rtl, noSideEffect, extern: "nos$1".} =
+      result = dir
+      continue
+    if dir == "" and name != "":
+      return "."
```
Really? What is this "." directory and if it is one why is ".." not a directory?

Araq commented on lib/pure/os.nim:
```diff
@@ -666,6 +589,82 @@ proc isAbsolute*(path: string): bool {.rtl, noSideEffect, extern: "nos$1".} =
+      continue
+    if dir == "" and name != "":
+      return "."
+    return dir
```
At this point in the control flow it's not obvious if it shouldn't have been ``return result``.

krux02 commented:
too long didn't read.

timotheecour commented:
instead of addressing remaining comments in this PR I'm opening https://github.com/nim-lang/Nim/pull/10633 which has a different, simpler implementation and should address the remaining comments from here; if #10633 gets merged I'll close this PR.

krux02 commented:
this I am closing this in favor of  #10633
